File: src/cmd/task.rs
use super::root::Context;
use crate::internal::{clierrors, config, generation};
use clap::{Args, Subcommand};

#[derive(Debug, Args)]
pub struct TaskCommand {
    #[command(subcommand)]
    command: TaskSubcommand,
}

#[derive(Debug, Subcommand)]
enum TaskSubcommand {
    #[command(about = "Get the current status of a generation task")]
    Status(TaskStatusArgs),
}

#[derive(Debug, Args)]
struct TaskStatusArgs {
    task_id: String,
    #[arg(
        long,
        default_value = "",
        help = "Output format: url (URLs only), json (full response)"
    )]
    output: String,
}

pub fn handle(cmd: TaskCommand, _ctx: Context) -> anyhow::Result<()> {
    match cmd.command {
        TaskSubcommand::Status(args) => status(args),
    }
}

fn status(args: TaskStatusArgs) -> anyhow::Result<()> {
    let cfg = config::load()?;
    if cfg.api_key.is_empty() {
        return Err(clierrors::err_no_api_key().into());
    }

    let task = generation::get_task(&cfg.api_key, &args.task_id)
        .map_err(|err| anyhow::anyhow!("failed to fetch task {}: {err}", args.task_id))?;

    match args.output.as_str() {
        "url" => {
            for url in task.urls() {
                println!("{url}");
            }
        }
        "json" => {
            println!("{}", serde_json::to_string_pretty(&task)?);
        }
        "" => {
            println!("Task:   {}", task.id);
            println!("Status: {}", task.status);
            if task.status == "failed" {
                if let Some(error) = &task.error {
                    println!("Error:  {}", error.message);
                }
            }
            for url in task.urls() {
                println!("URL:    {url}");
            }
        }
        _ => {}
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::internal::test_support::{TestResponse, TestServer};
    use serial_test::serial;
    use std::env;

    #[test]
    #[serial]
    fn status_command_covers_human_json_url_and_failed() {
        let server = TestServer::new(|req| {
            assert!(req.path.starts_with("/model/v1/generation/task/"));
            if req.path.ends_with("failed") {
                return TestResponse::json(
                    200,
                    r#"{"id":"failed","status":"failed","error":"bad"}"#,
                );
            }
            TestResponse::json(
                200,
                r#"{"id":"task-1","status":"completed","output":[{"content":[{"url":"https://cdn.example.com/out.png"}]}]}"#,
            )
        });
        env::set_var("SEACLOUD_GENERATION_URL", server.url());
        env::set_var(config::ENV_FOLKOS_EXEC_TOKEN, "managed-token");

        status(TaskStatusArgs {
            task_id: "task-1".to_string(),
            output: String::new(),
        })
        .unwrap();
        status(TaskStatusArgs {
            task_id: "task-1".to_string(),
            output: "json".to_string(),
        })
        .unwrap();
        status(TaskStatusArgs {
            task_id: "task-1".to_string(),
            output: "url".to_string(),
        })
        .unwrap();
        status(TaskStatusArgs {
            task_id: "failed".to_string(),
            output: String::new(),
        })
        .unwrap();
    }
}

File: cmd/task.go
package cmd

import (
	"encoding/json"
	"fmt"

	"github.com/SeaCloudAI/seacloud-cli/internal/clierrors"
	"github.com/SeaCloudAI/seacloud-cli/internal/config"
	"github.com/SeaCloudAI/seacloud-cli/internal/generation"
	"github.com/spf13/cobra"
)

var taskStatusOutput string

var taskCmd = &cobra.Command{
	Use:   "task",
	Short: "Check generation task status and outputs",
	Long: `Manage SeaCloud generation tasks.

Use this after an async model run returns a task ID.`,
}

var taskStatusCmd = &cobra.Command{
	Use:   "status <task_id>",
	Short: "Get the current status of a generation task",
	Long: `Fetch the current status of a generation task by its ID.

Exit codes:
  0   request succeeded (task may still be in_progress)
  1   error (network, API, missing credentials)`,
	Example: `  seacloud task status d758n65e878c73cmdg20
  seacloud task status d758n65e878c73cmdg20 --output url
  seacloud task status d758n65e878c73cmdg20 --output json`,
	Args: cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		taskID := args[0]

		cfg, err := config.Load()
		if err != nil {
			return err
		}
		if cfg.APIKey == "" {
			return clierrors.ErrNoAPIKey()
		}

		if task, handled, err := getQueueTaskStatus(cfg.APIKey, taskID); handled {
			if err != nil {
				return err
			}
			return printQueueTaskStatus(task)
		}

		task, err := generation.GetTask(cfg.APIKey, taskID)
		if err != nil {
			return fmt.Errorf("failed to fetch task %s: %w", taskID, err)
		}

		if taskStatusOutput == "url" {
			for _, u := range task.URLs() {
				fmt.Println(u)
			}
			return nil
		}

		if taskStatusOutput == "json" {
			b, _ := json.MarshalIndent(task, "", "  ")
			fmt.Println(string(b))
			return nil
		}

		// Human-readable
		fmt.Printf("Task:   %s\n", task.ID)
		fmt.Printf("Status: %s\n", task.Status)
		if task.Status == "failed" && task.Error != nil {
			fmt.Printf("Error:  %s\n", task.Error.Message)
		}
		for _, u := range task.URLs() {
			fmt.Printf("URL:    %s\n", u)
		}
		return nil
	},
}

func init() {
	taskStatusCmd.Flags().StringVar(&taskStatusOutput, "output", "", "Output format: url (URLs only), json (full response)")
	taskCmd.AddCommand(taskStatusCmd)
}
